### ISSUE TITLE
piping(?) freezes xonsh

### ISSUE THREAD

**Issue by laerus** [xonsh/xonsh#1330]
both less and more don't show any output before freeze
`XONSH_SHOW_TRACEBACK=True` doesn't show anything

```                                                                                                    
$ python3 -m pytest --help | less
$ jobs
[1]+ stopped: python3 -m pytest --help | less (19744)
$ fg
[1]+ running: python3 -m pytest --help | less (19744)

^C
```

```
$ cat random_HUGE_file.txt | more
$ fg
[1]+ running: cat random_HUGE_file.txt | more (20426)
^C^C
```

anyone can reproduce?

**Comment by scopatz** [comment]
Thanks for reporting @laerus!  I can reproduce, but I don't have a good sense of why this is happening.

**Comment by PaulReiber** [comment]
Same occurs if you pipe something to "view -"

I also tried this:
ls | /usr/games/rain

...and couldn't even get control-c's or control-\'s to kill rain - had to
abort my ssh session to kill it.
Controlling terminal access maybe?  Session?  Not sure - but there's
definitely a bug here.

Kind regards,


*Paul Reiber*http://about.me/reiber

**Comment by PaulReiber** [comment]
Hello xonsh-erers,

Happy to share results of further testing regarding pipes.

TL;DR is, xonsh's pipes don't seem to be answering as they should for a
pipeline. From the source subprocess's point of view, if stdout is a pipe,
sys.stdout.isatty() needs to return False.

pbr@mongo:~$ cat tty_test/idea

# comparing pipes test

in bash:
python -c "import sys; print(sys.stdout.isatty())" # should be True
python -c "import sys; print(sys.stdout.isatty())" | grep . # should be False

in xonsh:
python -c "import sys; print(sys.stdout.isatty())" # should be True
python -c "import sys; print(sys.stdout.isatty())" | grep . # should be False

...ok, so let's see what happens:
pbr@mongo:~$ python -c "import sys; print(sys.stdout.isatty())"
True
pbr@mongo:~$ python -c "import sys; print(sys.stdout.isatty())" |cat
False
pbr@mongo:~$ . xonsh/venv/bin/activate
(venv)pbr@mongo:~$ which xonsh
/home/pbr/xonsh/venv/bin/xonsh
(venv)pbr@mongo:~$ xonsh
(venv) pbr@mongo ~ $ import sys
(venv) pbr@mongo ~ $ print(sys.stdout.isatty())
True
(venv) pbr@mongo ~ $ print(sys.stdout.isatty()) | cat
True
xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
NameError: name 'cat' is not defined
(venv) pbr@mongo ~ $

So, thet last one should be returning a "False" there, not "True".

I noted then in testing that xonsh didn't seem to want to exit when
stdin had an end-of-file - control-D.

Eventually control-\ got me all the way out to bash again

re-running with XONSH_SHOW_TRACEBACK = True  for you... now we're
getting somewhere!

(venv)pbr@mongo:~$ xonsh
(venv) pbr@mongo ~ $ $XONSH_SHOW_TRACEBACK = True
(venv) pbr@mongo ~ $ import sys
(venv) pbr@mongo ~ $ $XONSH_SHOW_TRACEBACK = True
(venv) pbr@mongo ~ $ print(sys.stdout.isatty())
True
(venv) pbr@mongo ~ $ print(sys.stdout.isatty()) | cat
True
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE =
<filename>
Traceback (most recent call last):
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/__amalgam__.py",
line 12880, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/__amalgam__.py",
line 2817, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
NameError: name 'cat' is not defined
(venv) pbr@mongo ~ $ print(sys.stdout.isatty()) | /bin/cat -
....................
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE =
<filename>
Traceback (most recent call last):
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/__amalgam__.py",
line 15821, in _parse_ctx_free
    debug_level=(self.debug_level > 1))
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/parsers/base.py",
line 307, in parse
    tree = self.parser.parse(input=s, lexer=self.lexer, debug=debug_level)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/ply/yacc.py",
line 325, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/ply/yacc.py",
line 1175, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/ply/yacc.py",
line 187, in call_errorfunc
    r = errorfunc(token)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/parsers/base.py",
line 2340, in p_error
    self.xonsh_code)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/parsers/base.py",
line 408, in _parse_error
    raise err
  File "<string>", line None
SyntaxError: <xonsh-code>:1:27: ('code: ![',)
print(sys.stdout.isatty()) ![| /bin/cat -]
                           ^

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/__amalgam__.py",
line 12920, in push
    locs=None)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/__amalgam__.py",
line 15768, in compile
    tree = self.parse(input, ctx, mode=mode, transform=transform)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/__amalgam__.py",
line 15739, in parse
    tree, input = self._parse_ctx_free(input, mode=mode)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/__amalgam__.py",
line 15879, in _parse_ctx_free
    raise original_error
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/__amalgam__.py",
line 15821, in _parse_ctx_free
    debug_level=(self.debug_level > 1))
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/parsers/base.py",
line 307, in parse
    tree = self.parser.parse(input=s, lexer=self.lexer, debug=debug_level)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/ply/yacc.py",
line 325, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/ply/yacc.py",
line 1175, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/ply/yacc.py",
line 187, in call_errorfunc
    r = errorfunc(token)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/parsers/base.py",
line 2340, in p_error
    self.xonsh_code)
  File "/home/pbr/xonsh/venv/local/lib/python3.4/site-packages/xonsh/parsers/base.py",
line 408, in _parse_error
    raise err
  File "<string>", line None
SyntaxError: <xonsh-code>:1:29: ('code: /',)
print(sys.stdout.isatty()) | /bin/cat -
                             ^
(venv) pbr@mongo ~ $


Anthony - there's definitely a couple different things going on here.

If you have questions for me, follow-on tests, etc. just ask.


Kind regards,


*Paul Reiber*http://about.me/reiber

**Comment by adqm** [comment]
For the second example, we don't currently have a nice way of piping Python-mode commands to subprocess-mode commands (or _vice versa_), so `print(something) | some_command` will not work, but `echo something | some_command` should.  There may be an opportunity to make this work across modes, but it is non-trivial (since `print(something) | some_command` is valid Python using the "bitwise or" operator `|`).

The other issue (with `less` dying) is one I can confirm.  And it is somewhat of a mystery to me, but I have an aidea of what might be causing it, and I have some potential fixes in mind.

**Comment by PaulReiber** [comment]
Thanks Adam,

Yes, I had sussed out that the | was being interpreted in python mode not
subprocess mode.

(venv) pbr@mongo ~/xonsh master $ echo @(sys.stdout.isatty()) | cat
True

...should be returning False, not True - so the problem is still
there, it was just compounded with the other matter.



Kind regards,


*Paul Reiber*http://about.me/reiber

**Comment by wreed4** [comment]
I'm also having problems with pipes.  `echo "testing" | vim -` no longer works for me at all.  When vim would have opened, xonsh backgrounds it and leaves it as a dangling process.  I can't switch to it with `fg`, I get a `ProcessLookupError` even though, before I do `fg`, `jobs` shows I have a process running in the background.  This is an incredibly common workflow for me, so any work-around would be greatly appreciated.

**Comment by adqm** [comment]
Can you post the output of `xonfig`?  I believe this has been fixed on current master, though the fix is not in an official release yet.

**Comment by wreed4** [comment]
```
+------------------+---------------------------------+
| xonsh            | 0.4.2                           |
| Python           | 3.5.1                           |
| PLY              | 3.7                             |
| have readline    | True                            |
| prompt toolkit   | 1.0.3                           |
| shell type       | prompt_toolkit                  |
| pygments         | 2.1.3                           |
| on posix         | True                            |
| on linux         | True                            |
| distro           | Red Hat Enterprise Linux Server |
| on darwin        | False                           |
| on windows       | False                           |
| on cygwin        | False                           |
| is superuser     | False                           |
| default encoding | utf-8                           |
| git SHA          | None                            |
+------------------+---------------------------------+
```

**Comment by wreed4** [comment]
The new release of xonsh fixes this problem for me. (v 0.4.3)
